spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/rx/ObservableReturnValueHandler.java

```diff
@@ -17,55 +17,129 @@
+import rx.Observer;
+
+import java.io.IOException;
+import java.io.OutputStream;
+import java.util.List;
 
 /**
  * Handles return values of type {@link rx.Observable}.
  *
  * @author Spencer Gibb
+ * @author Jakub Narloch
  */
-public class ObservableReturnValueHandler implements AsyncHandlerMethodReturnValueHandler {
+public class ObservableReturnValueHandler extends AbstractWrappableReturnValueHandler<Observable> {
 
-	@Override
-	public boolean isAsyncReturnValue(Object returnValue, MethodParameter returnType) {
-		return returnValue != null && returnValue instanceof Observable;
+	private final List<HttpMessageConverter<?>> messageConverters;
+
+	public ObservableReturnValueHandler(List<HttpMessageConverter<?>> messageConverters) {
+		super(Observable.class);
+		Assert.notEmpty(messageConverters, "'messageConverters' must not be empty");
+		this.messageConverters = messageConverters;
 	}
 
 	@Override
-	public boolean supportsReturnType(MethodParameter returnType) {
-		return Observable.class.isAssignableFrom(returnType.getParameterType());
+	protected void handleUnwrappedReturnValue(Observable observable, MethodParameter methodParameter, ModelAndViewContainer mavContainer, NativeWebRequest webRequest, ServerHttpResponse response) throws Exception {
+
+		final DeferredResult<Object> deferredResult = new DeferredResult<>();
+		observable.subscribe(new HttpMessageConverterObserver(deferredResult, new StreamingServletServerHttpResponse(response)));
+		WebAsyncUtils.getAsyncManager(webRequest).startDeferredResultProcessing(deferredResult, mavContainer);
 	}
 
-	@Override
-	public void handleReturnValue(Object returnValue, MethodParameter returnType, ModelAndViewContainer mavContainer, NativeWebRequest webRequest) throws Exception {
-		if (returnValue == null) {
-			mavContainer.setRequestHandled(true);
-			return;
+	private class HttpMessageConverterObserver implements Observer {
+
+		private final DeferredResult deferredResult;
+		private final ServerHttpResponse response;
+		private boolean completed;
+
+		private HttpMessageConverterObserver(DeferredResult deferredResult, ServerHttpResponse response) {
+			this.deferredResult = deferredResult;
+			this.response = response;
 		}
 
-		Observable<?> observable = Observable.class.cast(returnValue);
+		@Override
+		public void onNext(Object next) {
+			if(!completed) {
+				try {
+					send(next);
+				} catch (IOException e) {
+					throw new RuntimeException("Unexpected error occurred during Observable streaming", e);
+				}
+			}
+		}
 
-		final DeferredResult<Object> deferredResult = new DeferredResult<>();
+		@Override
+		public void onCompleted() {
+			if(!completed) {
+				completed = true;
+				deferredResult.setResult(null);
+			}
+		}
+
+		@Override
+		public void onError(Throwable throwable) {
+			deferredResult.setErrorResult(throwable);
+		}
+
+		private <T> void send(T next) throws IOException {
+			final MediaType mediaType = response.getHeaders().getContentType();
```

Comment:
This is the only piece that might be a complication, the original ResponseBodyEmitter have information about the object media type, at this point I don't have anything similar from Observable.